core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java

```diff
@@ -107,47 +106,69 @@
   private BuiltInMetadata.LowerBoundCost.Handler lowerBoundCostHandler;
 
   /**
-   * Creates the instance with {@link JaninoRelMetadataProvider} instance
-   * from {@link #THREAD_PROVIDERS} and {@link #EMPTY} as a prototype.
+   * Creates the instance using the Handler.classualt prototype.
    */
   protected RelMetadataQuery() {
-    this(castNonNull(THREAD_PROVIDERS.get()), EMPTY);
+    this(PROTOTYPE);
   }
 
   /** Creates and initializes the instance that will serve as a prototype for
    * all other instances. */
-  private RelMetadataQuery(@SuppressWarnings("unused") boolean dummy) {
-    super(null);
-    this.collationHandler = initialHandler(BuiltInMetadata.Collation.Handler.class);
-    this.columnOriginHandler = initialHandler(BuiltInMetadata.ColumnOrigin.Handler.class);
-    this.expressionLineageHandler = initialHandler(BuiltInMetadata.ExpressionLineage.Handler.class);
-    this.tableReferencesHandler = initialHandler(BuiltInMetadata.TableReferences.Handler.class);
-    this.columnUniquenessHandler = initialHandler(BuiltInMetadata.ColumnUniqueness.Handler.class);
-    this.cumulativeCostHandler = initialHandler(BuiltInMetadata.CumulativeCost.Handler.class);
-    this.distinctRowCountHandler = initialHandler(BuiltInMetadata.DistinctRowCount.Handler.class);
-    this.distributionHandler = initialHandler(BuiltInMetadata.Distribution.Handler.class);
-    this.explainVisibilityHandler = initialHandler(BuiltInMetadata.ExplainVisibility.Handler.class);
-    this.maxRowCountHandler = initialHandler(BuiltInMetadata.MaxRowCount.Handler.class);
-    this.minRowCountHandler = initialHandler(BuiltInMetadata.MinRowCount.Handler.class);
-    this.memoryHandler = initialHandler(BuiltInMetadata.Memory.Handler.class);
-    this.nonCumulativeCostHandler = initialHandler(BuiltInMetadata.NonCumulativeCost.Handler.class);
-    this.parallelismHandler = initialHandler(BuiltInMetadata.Parallelism.Handler.class);
+  protected RelMetadataQuery(MetadataHandlerProvider metadataHandlerProvider) {
```

Comment:
At this point the MetadataProvider a list a of tuples (Class<? extends MetadataHandler>, MetadataHandler instance) which represent binding a set methods to a particular metadata.  Without this patch the Metadata class is not used in any meaningful way in code path that is not already deprecated.  You have to have a symbol(global constant) to bind a the metadata implementation too.

Part of the whole point of the interface is to let people customize/change the behavior of the handler generation.  Some people want to eagerly generate the code at runtime, other want to tweaks the caching of generated classes, you and others do not want to use janino altogether.  This would support all of those use cases.

RelMetadataQuery needs to call the implementation of the metadata calls which is currently done through handlers.  RelMetadataQueryBase already has a supported api for doing conceptually similar things with similar amount of exposed api.  Their are just a set of things that you can not do with that existing and supported API that I am attempting to rectify.

Several consumer of apache calcite have custom metadata calls.  Currently the suggested way to create a custom call is to subclass RelMetadataQuery.  What you are purposing would require 2 different subclasses of RelMetadataQuery and also require them to correctly maintain the caching/cyclic dependency boiler plate.  I understand the desire to be able to debug code  and have a separation of concerns but the API you are proposing would require several subclasses to do something conceptually simple and introduce breaking changes.

Reply:
To put this another way: RelMetadataQuery is the what, RelMetadataProvider is one possible how. Let's not couple the two.

Reply:
There are several requirements metadata in calcite:
Metadata retrieval - the api for getting a metdata value for a relnode and give set of arguments.  This is currently the RelMetadataQuery.
Metadata implementation - The code that is called to for a relnode to get a metadata value.  This is currently sub-classes of RelMetdataProvider.
Defaulting and Validation - There is defaulting, canonization, validation, etc after a metadata value is retrieved in RelMetadataQuery.
Dispatch - How the implementation is called for given arguments.  This is currently done in Janino compiled code.
Cycle Detection - This is currently a **hard** requirement of volcano planner.  This is currently done in Janino compiled code.
Caching Invalidation - This is required for volcano planner.  This is currently done through RelMetadataQuery.clearCache.
Caching Implementation - This is coupled with cycle detection, but is currently accomplished through RelMetadataQuery.map.
Caching Values - The caches need to be checked and updated.  This is currently done in the generated code.

Metadata retrieval and Metadata implementation are hard requirements for external facing APIs.  They are currently exist as external facing APIs.  Many downstream projects depend on the ability to either customize existing metadata types or add their own metadata types.  When a down stream project adds their own metadata type, they need to be able to expose it to the larger system aka through metadata retrieval.  This is currently done by subclassing RelMetadataQuery and calling methods RelMetadataQueryBase.  An example of this can be seen in RelMetadataTest.MyRelMetadataQuery.

So the metadata retrieval api is the only api need to for the rest of calcite core, however, downstream users of calcite need to be able to configure and extend the metadata implementation.  So some of RelMetadataQuery implementation must be exposed with the current setup.

I support making RelMetadataQuery an interface so the metadata retrieval is clearly defined.  However, this is not the only public API necessary to expose.  Also, it is breaking change which does not separate the defaulting and validation from the RelMetadataQueryImpl.  If RelMetadataQuery was split into RelMetadataQuery(an interface), DefaultingAndValidatingRelMetadataQuery and HandlerBackedRelMetadataQuery, then down stream projects would have to implement 2 different RelMetadataQueries to add a custom metadata.  This api seems a bit convoluted and cumbersome.  Alternatively, breaking it up into class hierarchy of RelMetadataQuery -> DefaultingAndValidatingRelMetadataQuery ->  HandlerBackedRelMetadataQuery also feels a bit convoluted and is a breaking change.

I also support removing the leaky abstraction of the metadata implementation to the rest of calcite core.  But changing this does not actually help things much other than remove a thread local that can cause odd behaviors in uncommon nesting scenarios.

Reply:
I think I have a simple solution to avoiding the introduction of a new constructor on this class:
- Make this constructor private
- Add a new inner static subclass that uses that constructor and exposes a protected class.

Then we tell people that want to use the handler behavior to extend that subclass. That way, we have deprecated subclassing RelMetadataQuery directly while moving towards this new pattern with minimal code changes.

This avoids the introduction of a new coupling api between RelMetadataQuery the handler concept, is minimal changes in your patch and sets us on a path to moving RelMetadataQuery to an interface or dumb abstract class.